### ISSUE TITLE
High Error rate in Classification

### ISSUE THREAD

**Issue by ambarpal** [AlfredXiangWu/face_verification_experiment#4]
Hi Alfred,
I was testing your pretrained model on CASIA-Webface itself. While evaluating the images under `CASIA-WebFace/0005282/` I get the following result: 
016.jpg predicted class: 8315
009.jpg predicted class: 6904
036.jpg predicted class: 8344
019.jpg predicted class: 1798
024.jpg predicted class: 1798
017.jpg predicted class: 4576
011.jpg predicted class: 1798
029.jpg predicted class: 10047
033.jpg predicted class: 6232
012.jpg predicted class: 1452
034.jpg predicted class: 6904
026.jpg predicted class: 1452
031.jpg predicted class: 1798
027.jpg predicted class: 6904
021.jpg predicted class: 9743
001.jpg predicted class: 1452
018.jpg predicted class: 6904
002.jpg predicted class: 8616
038.jpg predicted class: 1798
007.jpg predicted class: 1452
032.jpg predicted class: 9400
040.jpg predicted class: 10427
010.jpg predicted class: 10427
025.jpg predicted class: 1452
006.jpg predicted class: 9137
014.jpg predicted class: 8153
037.jpg predicted class: 6759
013.jpg predicted class: 10427
023.jpg predicted class: 1798
028.jpg predicted class: 9400
030.jpg predicted class: 1798
039.jpg predicted class: 1798
015.jpg predicted class: 6998
005.jpg predicted class: 10427
003.jpg predicted class: 10427
008.jpg predicted class: 9248
020.jpg predicted class: 6904
022.jpg predicted class: 7846
004.jpg predicted class: 2791
035.jpg predicted class: 6904
041.jpg predicted class: 1798

The predictions look very random to me. Here is my code that produced this result:

``` python
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.cm as cm
from os import walk

caffe_root = '/home/ambar/caffe/'
import sys
sys.path.insert(0, '/home/ambar/caffe/python')
import caffe

MODEL_FILE = '/home/ambar/casia-network/face_verification_experiment/proto/DeepFace_set003.prototxt'
PRETRAINED = '/home/ambar/casia-network/face_verification_experiment/model/DeepFace_set003_net_iter.caffemodel'

caffe.set_mode_gpu()
net = caffe.Classifier(model_file=MODEL_FILE, pretrained_file=PRETRAINED, raw_scale=128, image_dims=(128,128), mean=None)
def outputClassification(img, mode=0):
    if (mode == 1): caffe.set_mode_gpu()
    im = caffe.io.load_image(img, False)
    prediction = net.predict([im])
    print 'predicted class:', prediction[0].argmax()
def getclasses(base):
    for (dirpath, dirnames, filenames) in walk(base):
        for i in filenames:
            print i,
            outputClassification(dirpath + i, 1)
getclasses('/home/ambar/casia/CASIA-WebFace/CASIA-WebFace/0000204/gray/') #images converted to gray without normalisation
getclasses('/home/ambar/casia/Normalized_Faces/webface/100/0000204/') #normalised images
```

I then ran the same on the normalised dataset `Normalized_Faces/webface/100/0000204/`, and get the same results: Everytime the class `1452` has a lot of predictions, whereas the other classes seem random. This happens with no matter which celebrity faces I choose.
Please tell me where I am going wrong.

**Comment by AlfredXiangWu** [comment]
I think your program and testing results are right.

This is a part of log for training this model
I0422 11:08:57.297183  8128 solver.cpp:191] Iteration 1750700, loss = 1.88449
I0422 11:08:57.297245  8128 solver.cpp:206]     Train net output #0: accuracy = 0.75
I0422 11:08:57.297256  8128 solver.cpp:403] Iteration 1750700, lr = 0.0001
I0422 11:09:32.724076  8128 solver.cpp:191] Iteration 1750800, loss = 2.10125
I0422 11:09:32.724151  8128 solver.cpp:206]     Train net output #0: accuracy = 0.703125
I0422 11:09:32.724164  8128 solver.cpp:403] Iteration 1750800, lr = 0.0001
I0422 11:10:08.156371  8128 solver.cpp:191] Iteration 1750900, loss = 1.65417
I0422 11:10:08.156442  8128 solver.cpp:206]     Train net output #0: accuracy = 0.71875
I0422 11:10:08.156455  8128 solver.cpp:403] Iteration 1750900, lr = 0.0001
I0422 11:10:43.235502  8128 solver.cpp:247] Iteration 1751000, Testing net (#0)
I0422 11:10:46.454829  8128 solver.cpp:285] Test loss: 0.656418
I0422 11:10:46.454869  8128 solver.cpp:298]     Test net output #0: accuracy = 0.901
As is shown in log, the accuracy of validation set is 0.901

This is the testing results of 50000(iterations:50. batch_size:100 ) face images of training set via "caffe test".
I0716 20:22:06.411171 12785 net.cpp:505] Collecting Learning Rate and Weight Decay.
I0716 20:22:06.411201 12785 net.cpp:257] Network initialization done.
I0716 20:22:06.411208 12785 net.cpp:258] Memory required for data: 1431910008
I0716 20:22:06.448001 12785 caffe.cpp:146] Running for 50 iterations.
I0716 20:22:18.701198 12785 caffe.cpp:170] Batch 0, accuracy = 0.95
I0716 20:22:30.031234 12785 caffe.cpp:170] Batch 1, accuracy = 0.89
I0716 20:22:41.345909 12785 caffe.cpp:170] Batch 2, accuracy = 0.9
I0716 20:22:52.657845 12785 caffe.cpp:170] Batch 3, accuracy = 0.9
I0716 20:23:03.972290 12785 caffe.cpp:170] Batch 4, accuracy = 0.89
I0716 20:23:15.288380 12785 caffe.cpp:170] Batch 5, accuracy = 0.86
I0716 20:23:26.602849 12785 caffe.cpp:170] Batch 6, accuracy = 0.88
I0716 20:23:37.918858 12785 caffe.cpp:170] Batch 7, accuracy = 0.91
I0716 20:23:49.234144 12785 caffe.cpp:170] Batch 8, accuracy = 0.88
I0716 20:24:00.554193 12785 caffe.cpp:170] Batch 9, accuracy = 0.9
I0716 20:24:11.870033 12785 caffe.cpp:170] Batch 10, accuracy = 0.91
I0716 20:24:23.182193 12785 caffe.cpp:170] Batch 11, accuracy = 0.92
I0716 20:24:34.495973 12785 caffe.cpp:170] Batch 12, accuracy = 0.9
I0716 20:24:45.809303 12785 caffe.cpp:170] Batch 13, accuracy = 0.91
I0716 20:24:57.124562 12785 caffe.cpp:170] Batch 14, accuracy = 0.86
I0716 20:25:08.439096 12785 caffe.cpp:170] Batch 15, accuracy = 0.92
I0716 20:25:19.747367 12785 caffe.cpp:170] Batch 16, accuracy = 0.89
I0716 20:25:31.060384 12785 caffe.cpp:170] Batch 17, accuracy = 0.88
I0716 20:25:42.371222 12785 caffe.cpp:170] Batch 18, accuracy = 0.92
I0716 20:25:53.685411 12785 caffe.cpp:170] Batch 19, accuracy = 0.82
I0716 20:26:04.995203 12785 caffe.cpp:170] Batch 20, accuracy = 0.92
I0716 20:26:16.309063 12785 caffe.cpp:170] Batch 21, accuracy = 0.93
I0716 20:26:27.619685 12785 caffe.cpp:170] Batch 22, accuracy = 0.9
I0716 20:26:38.933145 12785 caffe.cpp:170] Batch 23, accuracy = 0.88
I0716 20:26:50.245322 12785 caffe.cpp:170] Batch 24, accuracy = 0.91
I0716 20:27:01.559742 12785 caffe.cpp:170] Batch 25, accuracy = 0.89
I0716 20:27:12.869074 12785 caffe.cpp:170] Batch 26, accuracy = 0.91
I0716 20:27:24.183049 12785 caffe.cpp:170] Batch 27, accuracy = 0.89
I0716 20:27:35.498638 12785 caffe.cpp:170] Batch 28, accuracy = 0.89
I0716 20:27:46.812515 12785 caffe.cpp:170] Batch 29, accuracy = 0.86
I0716 20:27:58.126798 12785 caffe.cpp:170] Batch 30, accuracy = 0.89
I0716 20:28:09.439180 12785 caffe.cpp:170] Batch 31, accuracy = 0.87
I0716 20:28:20.748698 12785 caffe.cpp:170] Batch 32, accuracy = 0.92
I0716 20:28:32.063278 12785 caffe.cpp:170] Batch 33, accuracy = 0.9
I0716 20:28:43.378821 12785 caffe.cpp:170] Batch 34, accuracy = 0.94
I0716 20:28:54.703855 12785 caffe.cpp:170] Batch 35, accuracy = 0.91
I0716 20:29:06.017079 12785 caffe.cpp:170] Batch 36, accuracy = 0.91
I0716 20:29:17.339248 12785 caffe.cpp:170] Batch 37, accuracy = 0.86
I0716 20:29:28.646975 12785 caffe.cpp:170] Batch 38, accuracy = 0.9
I0716 20:29:39.951875 12785 caffe.cpp:170] Batch 39, accuracy = 0.92
I0716 20:29:51.256526 12785 caffe.cpp:170] Batch 40, accuracy = 0.91
I0716 20:30:02.571938 12785 caffe.cpp:170] Batch 41, accuracy = 0.95
I0716 20:30:13.879290 12785 caffe.cpp:170] Batch 42, accuracy = 0.89
I0716 20:30:25.186650 12785 caffe.cpp:170] Batch 43, accuracy = 0.92
I0716 20:30:36.503211 12785 caffe.cpp:170] Batch 44, accuracy = 0.89
I0716 20:30:47.812701 12785 caffe.cpp:170] Batch 45, accuracy = 0.93
I0716 20:30:59.120633 12785 caffe.cpp:170] Batch 46, accuracy = 0.95
I0716 20:31:10.441679 12785 caffe.cpp:170] Batch 47, accuracy = 0.84
I0716 20:31:21.751219 12785 caffe.cpp:170] Batch 48, accuracy = 0.91
I0716 20:31:33.066895 12785 caffe.cpp:170] Batch 49, accuracy = 0.95
I0716 20:31:33.066931 12785 caffe.cpp:175] Loss: 0.723969
I0716 20:31:33.066938 12785 caffe.cpp:187] accuracy = 0.9006

The reason for the performance may be as following:
1. The normalization of WebFace images is different. 
This is the original image of WebFace dataset
![original image](http://i.imgur.com/Xgm3giW.jpg)
This is the normalized image of our training set
![normalized image](http://i.imgur.com/NGDYy1m.jpg)

2.According to the face verification task based on CNN, it is certain that we often extract the first fully connected layer to represent the face image and the parameters from the first fully connected layer to the second one(label layer) is not used in testing. Therefore, while training the CNN model, I set weight_decay of fc2 to 10, which means L2 regularization, to avoid overfitting of training set.
Moreover, the performance of the classification for WebFace is not accurate, which cannot indicate the feature vector of face image is bad. For example, when the model is trained at iteration 40k, the validation set accuracy is 0.28, but the eer of lfw is 0.8763.

**Comment by ambarpal** [comment]
So, am I correct in understanding that if I now fine tune the network on a large number of images of a few faces, the accuracy for those faces will be appropriately high?

**Comment by AlfredXiangWu** [comment]
"if I now fine tune the network on a large number of images of a few faces, the accuracy for those faces will be appropriately high?"
I think so.
If you have any problem of training or finetuning the network, welcome to contact me.

**Comment by ambarpal** [comment]
Alright, I will. Thanks a lot for helping me.

**Comment by iscas-lee** [comment]
![20170208101346](https://cloud.githubusercontent.com/assets/7486566/22720707/54949de8-ede7-11e6-81df-2ade2a04ccff.jpg)
why don't you use pictures like this? The backgroud around the face is not useful. Is that so?
